app/src/main/scala/org/alephium/explorer/persistence/queries/InputsQueries.scala

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018 The Alephium Authors
+// This file is part of the alephium project.
+//
+// The library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the library. If not, see <http://www.gnu.org/licenses/>.
+
+package org.alephium.explorer.persistence.queries
+
+import scala.concurrent.ExecutionContext
+
+import com.typesafe.scalalogging.StrictLogging
+import slick.basic.DatabaseConfig
+import slick.dbio.DBIOAction
+import slick.jdbc.JdbcProfile
+
+import org.alephium.explorer.Hash
+import org.alephium.explorer.api.model._
+import org.alephium.explorer.persistence._
+import org.alephium.explorer.persistence.model._
+import org.alephium.explorer.persistence.schema._
+import org.alephium.util.U256
+
+trait InputsQueries extends InputSchema with OutputSchema with StrictLogging {
+
+  implicit def executionContext: ExecutionContext
+  val config: DatabaseConfig[JdbcProfile]
+  import config.profile.api._
+
+  private val mainInputs  = inputsTable.filter(_.mainChain)
+  private val mainOutputs = outputsTable.filter(_.mainChain)
+
+  def insertTxPerAddressFromInputs(inputs: Seq[InputEntity],
+                                   outputs: Seq[OutputEntity]): DBActionW[Seq[InputEntity]] = {
+
+    val inputsWithMaybeAddress = inputs.map(in => (in.address, in))
+
+    val assets = inputsWithMaybeAddress
+      .collect {
+        case (Some(address), input)
+            if !outputs.exists(out => out.address == address && out.txHash == input.txHash) =>
```

Comment:
This would be insanely slow when there are many inputs and outputs in the one block. The complexity is O(inputs.length * outputs.length).

Reply:
I improved [here](https://github.com/alephium/explorer-backend/pull/149/commits/d3fdfa7f7d38b19c0d70a23641276194c7a992e5) 

`Map.get` is O(1) right?